sdk/helper/docker/testhelpers.go

```diff
@@ -148,38 +179,112 @@ func (d *Runner) StartService(ctx context.Context, connect ServiceAdapter) (*Ser
+	scanner := bufio.NewScanner(bytes.NewReader(p))
+	scanner.Buffer(make([]byte, 64*1024), bufio.MaxScanTokenSize)
+	for scanner.Scan() {
+		l.consumer(scanner.Text())
+	}
+	return len(p), nil
+}
+
+var _ io.Writer = &LogConsumerWriter{}
+
 // StartNewService will start the runner's configured docker container but with the
 // ability to control adding a name suffix or forcing a local address to be returned.
 // 'addSuffix' will add a random UUID to the end of the container name.
 // 'forceLocalAddr' will force the container address returned to be in the
 // form of '127.0.0.1:1234' where 1234 is the mapped container port.
 func (d *Runner) StartNewService(ctx context.Context, addSuffix, forceLocalAddr bool, connect ServiceAdapter) (*Service, string, error) {
-	container, hostIPs, containerID, err := d.Start(context.Background(), addSuffix, forceLocalAddr)
+	if d.RunOptions.PreDelete {
+		name := d.RunOptions.ContainerName
+		matches, err := d.DockerAPI.ContainerList(ctx, types.ContainerListOptions{
+			All: true,
+			// TODO use labels to ensure we don't delete anything we shouldn't
+			Filters: filters.NewArgs(
+				filters.Arg("name", name),
+			),
+		})
+		if err != nil {
+			return nil, "", fmt.Errorf("failed to list containers named %q", name)
+		}
+		for _, cont := range matches {
+			err = d.DockerAPI.ContainerRemove(ctx, cont.ID, types.ContainerRemoveOptions{Force: true})
+			if err != nil {
+				return nil, "", fmt.Errorf("failed to pre-delete container named %q", name)
+			}
+		}
+	}
+	result, err := d.Start(context.Background(), addSuffix, forceLocalAddr)
 	if err != nil {
 		return nil, "", err
 	}
 
-	cleanup := func() {
-		if d.RunOptions.LogConsumer != nil {
-			rc, err := d.DockerAPI.ContainerLogs(ctx, container.ID, types.ContainerLogsOptions{
+	var wg sync.WaitGroup
+	consumeLogs := false
+	var logStdout, logStderr io.Writer
+	if d.RunOptions.LogStdout != nil && d.RunOptions.LogStderr != nil {
+		consumeLogs = true
+		logStdout = d.RunOptions.LogStdout
+		logStderr = d.RunOptions.LogStderr
+	} else if d.RunOptions.LogConsumer != nil {
+		consumeLogs = true
+		logStdout = &LogConsumerWriter{d.RunOptions.LogConsumer}
+		logStderr = &LogConsumerWriter{d.RunOptions.LogConsumer}
+	}
+
+	// The waitgroup wg is used here to support some stuff in NewDockerCluster.
+	// We can't generate the PKI cert for the https listener until we know the
+	// container's address, meaning we must first start the container, then
+	// generate the cert, then copy it into the container, then signal Vault
+	// to reload its config/certs.  However, if we SIGHUP Vault before Vault
+	// has installed its signal handler, that will kill Vault, since the default
+	// behaviour for HUP is termination.  So the PostStart that NewDockerCluster
+	// passes in (which does all that PKI cert stuff) waits to see output from
+	// Vault on stdout/stderr before it sends the signal, and we don't want to
+	// run the PostStart until we've hooked into the docker logs.
```

Comment:
That's slick and really handy!

Reply:
Haha I'm glad you think it's slick.  I agree it's handy but it feels like a messy kludge... still, it works.